walkeeper/src/bin/safekeeper.rs

```diff
@@ -123,6 +127,17 @@ fn start_safekeeper(conf: SafeKeeperConf) -> Result<()> {
 
     info!("version: {}", GIT_VERSION);
 
+    // Prevent running multiple safekeepers on the same directory
+    let lock_file_path = conf.workdir.join(LOCK_FILE_NAME);
+    let lock_file = File::create(&lock_file_path).with_context(|| "failed to open lockfile")?;
+    lock_file.try_lock_exclusive().map_err(|e| {
+        anyhow!(
+            "control file {:?} is locked by some other process: {}",
+            &lock_file_path,
```

Comment:
Few optional style nits:

* you can format `lock_file_path.display()` with `{}`, in case you prefer better control over what's formatting
* you can chain any standard error, so instead of `map_err` you can use `.context` and `.with_context` from `anyhow::Context`. This way, you get the error "hierarchy" the way anyhow provides it.